shared/src/shared.rs

```diff
@@ -95,77 +107,86 @@ impl<CI: ChainIndex> Shared<CI> {
                 }
             };
 
-            let output_root = match store.get_output_root(&header.hash()) {
-                Some(h) => h,
-                None => H256::zero(),
-            };
+            let txo_set = Self::init_txo_set(&store, header.hash());
 
             let total_difficulty = store
                 .get_block_ext(&header.hash())
                 .expect("block_ext stored")
                 .total_difficulty;
 
-            Arc::new(RwLock::new(TipHeader::new(
+            Arc::new(RwLock::new(ChainState::new(
                 header,
                 total_difficulty,
-                output_root,
+                txo_set,
             )))
         };
 
         Shared {
             store: Arc::new(store),
-            tip_header,
-            consensus,
+            chain_state,
+            consensus: Arc::new(consensus),
         }
     }
 
-    pub fn tip_header(&self) -> &RwLock<TipHeader> {
-        &self.tip_header
+    pub fn chain_state(&self) -> &RwLock<ChainState> {
+        &self.chain_state
     }
 
     pub fn store(&self) -> &Arc<CI> {
         &self.store
     }
-}
 
-impl<CI: ChainIndex> CellProvider for Shared<CI> {
-    fn cell(&self, out_point: &OutPoint) -> CellStatus {
-        let index = out_point.index as usize;
-        let tip_header = self.tip_header().read();
-        if let Some(meta) = self.get_transaction_meta(&tip_header.output_root, &out_point.hash) {
-            if index < meta.len() {
-                if !meta.is_spent(index) {
-                    let transaction = self
-                        .store
-                        .get_transaction(&out_point.hash)
-                        .expect("transaction must exist");
-                    CellStatus::Live(transaction.outputs()[index].clone())
-                } else {
-                    CellStatus::Dead
+    pub fn init_txo_set(store: &CI, mut hash: H256) -> TxoSet {
+        let mut txo_set = TxoSet::new();
+        let mut blocks = Vec::new();
+        while let Some(block) = store.get_block(&hash) {
+            let number = block.header().number();
+            hash = block.header().parent_hash().clone();
+            blocks.push(block);
```

Comment:
need to resolve memory issue here, e.g. it will consume 10G memory when restarting a node with 10,000 * 1M block.

Reply:
Good catch

Reply:
fixed